micro_sam/sam_annotator/_widgets.py

```diff
@@ -0,0 +1,56 @@
+from enum import Enum
+from pathlib import Path
+from typing import TYPE_CHECKING, Optional
+
+from magicgui import magic_factory
+from magicgui.tqdm import tqdm
+from superqt.utils import thread_worker
+
+from micro_sam.util import (
+    ImageEmbeddings,
+    get_sam_model,
+    precompute_image_embeddings,
+    _MODEL_URLS,
+    _DEFAULT_MODEL,
+)
+
+if TYPE_CHECKING:
+    import napari
+
+Model = Enum("Model", _MODEL_URLS)
+
+
+@magic_factory(call_button="Compute image embeddings",
+               device = {"choices": ["auto", 'cuda', 'mps', 'cpu']})
+def embedding_widget(
+    image: "napari.layers.Image",
+    model: Model = Model.__getitem__(_DEFAULT_MODEL),
+    device = "auto",
+    save_path: Optional[Path] = None,  # where embeddings for this image are cached (optional)
+    custom_model: Optional[str] = None,  # A filepath or URL to custom model weights.
+) -> ImageEmbeddings:
+    """Image embedding widget."""
+    # for access to the predictor and the image embeddings in the widgets
+    global PREDICTOR, IMAGE_EMBEDDINGS
+    # Initialize the model
+    PREDICTOR = get_sam_model(model_type=model.name)
+
+    # Get image dimensions
+    if not image.rgb:
+        ndim = image.data.ndim
+    else:
+        # assumes RGB channels are the last dimension
+        ndim = image.data.ndim
```

Comment:
Constantin said [here](https://github.com/computational-cell-analytics/micro-sam/issues/203#issue-1908560489) that the (spatial) image dimensions (`ndim`) should be saved to the global state. Is that true? 

I thought it seemed fine to just pass `ndim` into the function computing the embeddings, do we need it later on as well? It's possible I missed where that global variable is defined, can you clarify?

Reply:
This is potentially related to the issue you see with RGB. I will test this branch and follow up.

Reply:
I think you can ignore [the comment](https://github.com/computational-cell-analytics/micro-sam/issues/203#issue-1908560489).
However, the code you have written here will give the wrong result (and I think explains the long runtime you observe with RGB data you observe below).
The `ndim` variable must be equal to the **number of spatial dimensions**. So in the case of an RGB image it must be two, but your code yields three.
I would suggest to change this code to:
```python
ndim = image.data.ndim
if image.rgb:
   ndim -= 1
```
(This would potentially even work for 3D RGB data. Which is something I have not tried, so not sure if it will work in the code that computes embeddings, but let's not worry about that now).

Reply:
... damn, I can't believe I wrote this whole if/else statement out and FORGOT THE MINUS ONE. I'll fix it now 😆